packages/vscode-extension/src/debug/taskTerminal/localTunnelTaskTerminal.ts

```diff
@@ -0,0 +1,265 @@
+import { BaseTaskTerminal } from "./baseTaskTerminal";
+import { DepsManager, DepsType, LocalEnvManager } from "@microsoft/teamsfx-core";
+import { vscodeLogger } from "../depsChecker/vscodeLogger";
+import { vscodeTelemetry } from "../depsChecker/vscodeTelemetry";
+import { openTerminalCommand } from "../constants";
+import VsCodeLogInstance from "../../commonlib/log";
+
+const ngrokTunnelName = "bot";
+const ngrokEndpointRegex = /obj=tunnels name=bot addr=(?<src>.*) url=(?<endpoint>.*)/;
+const defaultNgrokBinFolderPlaceholder = "${teamsfx:ngrokBinFolder}";
+
+type LocalTunnelTaskStatus = {
+  resolvedConfigFile?: string;
+  endpoint?: string;
+  terminal: LocalTunnelTaskTerminal;
+};
+
+export interface LocalTunnelArgs {
+  configFile?: string;
+  binFolder?: string;
+  // TODO: reuse?: boolean
+}
+
+export class LocalTunnelTaskTerminal extends BaseTaskTerminal {
+  private static ngrokTaskTerminals: Map<string, LocalTunnelTaskStatus> = new Map<
+    string,
+    LocalTunnelTaskStatus
+  >();
+
+  private childProc: cp.ChildProcess | undefined;
+  private readonly taskTerminalId: string;
+  private readonly args: LocalTunnelArgs;
+  private readonly status: LocalTunnelTaskStatus;
+
+  constructor(taskDefinition: vscode.TaskDefinition) {
+    super(taskDefinition);
+    this.args = taskDefinition.args as LocalTunnelArgs;
+    this.taskTerminalId = uuidv4();
+
+    for (const task of LocalTunnelTaskTerminal.ngrokTaskTerminals.values()) {
+      task.terminal.close();
+    }
+
+    this.status = { terminal: this };
+    LocalTunnelTaskTerminal.ngrokTaskTerminals.set(this.taskTerminalId, this.status);
+  }
+
+  stop(error?: any): void {
+    this.childProc?.kill();
+    super.stop(error);
+    LocalTunnelTaskTerminal.ngrokTaskTerminals.delete(this.taskTerminalId);
+  }
+
+  do(): Promise<Result<Void, FxError>> {
+    return this.resolveArgs().then((v) => this.startNgrokChildProcess(v.configFile, v.binFolder));
+  }
+
+  private async resolveArgs(): Promise<{ configFile: string; binFolder?: string }> {
+    if (!this.args.configFile) {
+      throw BaseTaskTerminal.taskDefinitionError("configFile");
+    }
+
+    const configFile = BaseTaskTerminal.resolveTeamsFxVariables(this.args.configFile);
+    this.status.resolvedConfigFile = configFile;
+
+    const binFolder = this.args.binFolder
+      ? await LocalTunnelTaskTerminal.resolveBinFolder(
+          BaseTaskTerminal.resolveTeamsFxVariables(this.args.binFolder)
+        )
+      : undefined;
+
+    return { configFile: configFile, binFolder: binFolder };
+  }
+
+  private startNgrokChildProcess(
+    configFile: string,
+    binFolder?: string
+  ): Promise<Result<Void, FxError>> {
+    return new Promise<Result<Void, FxError>>((resolve, reject) => {
+      VsCodeLogInstance.info("Starting local tunnel task.");
+      const command = `ngrok start ${ngrokTunnelName} --config=${configFile} --log=stdout --log-format=logfmt`;
+      this.writeEmitter.fire(`${command}\r\n\r\n`);
```

Comment:
Does `\r\n` work on all OS platforms? Or use `os.EOL` instead?

Reply:
It is according to the vscode API document https://code.visualstudio.com/api/references/vscode-api#Pseudoterminal
![image](https://user-images.githubusercontent.com/49138419/188775671-f6a26eff-19c0-426e-be67-3cc29ccfdb77.png)

Reply:
Good to know!